subsys/bluetooth/services/cgms/cgms.c

```diff
@@ -0,0 +1,414 @@
+static ssize_t read_feature(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				void *buf, uint16_t len, uint16_t offset);
+static ssize_t read_status(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				void *buf, uint16_t len, uint16_t offset);
+static ssize_t read_session_start_time(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				void *buf, uint16_t len, uint16_t offset);
+static ssize_t write_session_start_time(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *buf, uint16_t len, uint16_t offset, uint8_t flags);
+static ssize_t read_session_run_time(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				void *buf, uint16_t len, uint16_t offset);
+static ssize_t racp_on_receive(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *buf, uint16_t len, uint16_t offset, uint8_t flags);
+static ssize_t socp_on_receive(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+				const void *buf, uint16_t len, uint16_t offset, uint8_t flags);
+
+BT_GATT_SERVICE_DEFINE(cgms_svc,
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_CGMS),
+	/*CGM Measurement*/
+	BT_GATT_CHARACTERISTIC(BT_UUID_CGM_MEASUREMENT, BT_GATT_CHRC_NOTIFY,
+				BT_GATT_PERM_READ_AUTHEN, NULL, NULL, NULL),
+	/* CGM Measurement CCCD */
+	BT_GATT_CCC(cgmc_ccc_cfg_changed,
+				BT_GATT_PERM_READ_AUTHEN | BT_GATT_PERM_WRITE_AUTHEN),
+	/* CGM Feature */
+	BT_GATT_CHARACTERISTIC(BT_UUID_CGM_FEATURE, BT_GATT_CHRC_READ,
+				BT_GATT_PERM_READ_AUTHEN,
+				read_feature, NULL, NULL),
+	/* CGM Status */
+	BT_GATT_CHARACTERISTIC(BT_UUID_CGM_STATUS, BT_GATT_CHRC_READ,
+				BT_GATT_PERM_READ_AUTHEN,
+				read_status, NULL, NULL),
+	/* CGM Session Start Time */
+	BT_GATT_CHARACTERISTIC(BT_UUID_CGM_SESSION_START_TIME,
+				BT_GATT_CHRC_READ  | BT_GATT_CHRC_WRITE,
+				BT_GATT_PERM_READ_AUTHEN | BT_GATT_PERM_WRITE_AUTHEN,
+				read_session_start_time, write_session_start_time, NULL),
+	/* CGM Session Run Time */
+	BT_GATT_CHARACTERISTIC(BT_UUID_CGM_SESSION_RUN_TIME, BT_GATT_CHRC_READ,
+				BT_GATT_PERM_READ_AUTHEN,
+				read_session_run_time, NULL, NULL),
+	/* Record Access Control Point */
+	BT_GATT_CHARACTERISTIC(BT_UUID_RECORD_ACCESS_CONTROL_POINT,
+				BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE,
+				BT_GATT_PERM_WRITE_AUTHEN,
+				NULL, racp_on_receive, NULL),
+	/* Record Access Control Point CCCD */
+	BT_GATT_CCC(cgmc_ccc_cfg_changed,
+				BT_GATT_PERM_READ_AUTHEN | BT_GATT_PERM_WRITE_AUTHEN),
+	/* CGM Specific Ops Control Point */
+	BT_GATT_CHARACTERISTIC(BT_UUID_CGM_SPECIFIC_OPS_CONTROL_POINT,
+				BT_GATT_CHRC_WRITE | BT_GATT_CHRC_INDICATE,
+				BT_GATT_PERM_WRITE_AUTHEN,
+				NULL, socp_on_receive, NULL),
+	/* CGM Specific Ops Control Point CCCD */
+	BT_GATT_CCC(cgmc_ccc_cfg_changed,
+				BT_GATT_PERM_READ_AUTHEN | BT_GATT_PERM_WRITE_AUTHEN),
+);
+
+static void cgmc_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	bool notif_enabled = (value != 0);
```

Comment:
As this ccc callback is shared by both measurement, which uses notification, and racp and socp, which uses indication, I need to catch both cases. So as long as it is non-zero (BT_GATT_CCC_NOTIFY = 1, BT_GATT_CCC_INDICATE = 2), it should imply the notification/indication is enabled.

Reply:
Right. It is possible to check notification and indirection independently.
I wonder if it would be a useful to add events about this for the user application.

Reply:
I prefer to keep it, as the developer of client-side app can use it for debugging.

Reply:
The message is also printed for indication.